docs/Implementing-easyblocks.rst

```diff
@@ -0,0 +1,486 @@
+  'EB_netCDF_minus_Fortran'
+
+Class name for generic easyblocks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For *generic* easyblocks, the class name does *not* include an ``EB_`` prefix, and the name is free to choose,
+taking into account the restriction to alphanumeric characters and underscores.
+
+For code style reasons, the class name should start with a capital letter.
+
+Examples include ``Bundle``, ``ConfigureMake``, ``CMakePythonPackage``.
+
+
+Module name/location
+~~~~~~~~~~~~~~~~~~~~
+
+The *name* of the Python module file is directly related to the name of Python class (i.e., the actual easyblock) that it provides.
+
+It should:
+
+* *not* include the ``EB_`` prefix of the class name for software-specific easyblocks
+* consists only of lower-case alphanumeric characters (``[a-z0-9]``) and underscores (``_``)
+
+  * dashes (``-``) are replaced by underscores (``_``)
+  * any other non-alphanumeric characters (incl. spaces) are simply dropped
+
+Examples include:
+
+* ``gcc.py`` (for *GCC*)
+* ``netcdf_fortran.py`` (for *netCDF-Fortran*)
+* ``gamess_us.py`` (for *GAMESS (US)*)
+
+The ``get_module_path`` function provided in ``easybuild.framework.easyconfig.easyconfig`` returns the (full)
+module location for a particular software name or easyblock class name; for example:
+
+.. code::
+
+  >>> from easybuild.framework.easyconfig.easyconfig import get_module_path
+  >>> get_module_path('netCDF-Fortran')
+  'easybuild.easyblocks.netcdf_fortran'
+  >>> get_module_path('EB_netCDF_minus_Fortran')
+  'easybuild.easyblocks.netcdf_fortran'
+
+The location of the Python module is determined by whether the easyblock is generic or software-specific.
+Generic easyblocks are located in the ``easybuid.easyblocks.generic`` namespace, while software-specific easyblocks
+live in the ``easybuild.easyblocks`` namespace directly. To keep things organised, the actual Python module file
+for software-specific easyblocks are kept in 'letter' subdirectories,
+rather than in one large '``easyblocks``' directory
+(see https://github.com/hpcugent/easybuild-easyblocks/blob/master/easybuild/easyblocks/).
+
+Note that you shouldn't concern yourself too much with getting the location of an easyblock right, as long as you
+use ``--include-easyblocks`` to make EasyBuild use additional or customised easyblocks;
+see :ref:`include_easyblocks` for more information.
+
+
+.. _implementing_easyblocks_structure:
+
+Structure of an easyblock
+-------------------------
+
+The example below shows the overal structure of an easyblock:
+
+.. code:: python
+
+    from easybuild.framework.easyblock import EasyBlock
+    from easybuild.tools.filetools import run_cmd
+
+    class EB_Example(EasyBlock):
+        """Custom easyblock for Example"""
+
+        def configure_step(self):
+            """Custom implementation of configure step for Example"""
+
+            # run configure.sh to configure the build
+            run_cmd("./configure.sh PREFIX=%s" % self.installdir)
```

Comment:
I'd be worried about some people seeing this and thinking "well, I can shove everything I want on `run_cmd`" and that's it. We actually have an easyblock implemented by somebody else in our institute that does a lot of that stuff. I'd look for a more pythonic example.